### PR TITLE
make: update stack-analysis format

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

Two things:

1. Makes the `make allstack` easier to read:

```
$ make stack-analysis
imix
----------------------
   main stack frame:
         2048     main

   5 largest stack frames:
         2048     main
          720     _ZN104_$LT$capsules..process_console..ProcessConsole$LT$C$GT$$u20$as$u20$kernel..hil..uart..TransmitClient$GT$18transmitted_buffer17h4c3fef6f6a95d58fE
          632     _ZN8capsules15process_console23ProcessConsole$LT$C$GT$11write_state17hee14a9c17699800aE
          608     _ZN7cortexm24kernel_hardfault_arm_v7m17h8c7b48b27a33c005E
          600     _ZN98_$LT$capsules..ieee802154..framer..Framer$LT$M$C$A$GT$$u20$as$u20$kernel..hil..radio..RxClient$GT$7receive17h8a8322dd16823f26E

   **WARNING! main is the largest stack frame!**
```

2. Enables `make stack-analysis V=1` to work fully.

### Testing Strategy

Running `make stack-analysis`.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: tools/stack_analysis.sh [hunk 1]
```diff
@@ -4,13 +4,28 @@
 # size information, and is thus most easily called using the `make stack-analysis`
 # rule.
 
-# Print the stack frame size of `main`
-printf "main stack frame: \n"
-$(find $(rustc --print sysroot) -name llvm-readobj) --elf-output-style GNU --stack-sizes $1 | grep 'main'
+bold=$(tput bold)
+normal=$(tput sgr0)
```
Comment (lschuermann) [inline tools/stack_analysis.sh:8]: Just a heads-up this adds a dependency on `ncurses`. It's shipped with my distro and seems to be on many, but it's worth keeping in mind - for instance I'm not quite sure whether Cygwin would have it by default.

**Review by lschuermann — APPROVED** [review]
Pretty cool.

File: tools/stack_analysis.sh (hunk 1 above)
Comment (ppannuto) [reply tools/stack_analysis.sh:8]: We've had `tput` around in the tools and build system for a long time; not a new dependency here, so I think it's fine.

File: tools/stack_analysis.sh [hunk 2]
```diff
@@ -4,13 +4,28 @@
...
+echo "$frames" | sort -n -r | head -5
 printf "\n"
+
+# Check if main is the largest stack frame
+largest=`echo "$frames" | sort -n -r | head -1 | grep ' main'`
+largest_ret_val=$?
+
+# If it is, print a warning.
+if [ $largest_ret_val -eq 0 ]; then
+    echo "   ${bold}WARNING! main is the largest stack frame!${normal}"
```
Comment (ppannuto) [inline tools/stack_analysis.sh:29]: Why is this something we warn about?

**Review by ppannuto — APPROVED** [review]

File: tools/stack_analysis.sh (hunk 2 above)
Comment (hudson-ayers) [reply tools/stack_analysis.sh:29]: That stack frame is never dropped

**Review by hudson-ayers — APPROVED** [review]

**Comment by phil-levis** [comment]
We need to be more careful and cautious about changing the output of tools. It can be that other tools depend on their output, such that changing it breaks those tools. This is an example of a churn/incremental improvement change which can be frustrating to people downstream.

**Review by alistair23 — APPROVED** [review]

File: tools/stack_analysis.sh (hunk 2 above)
Comment (ppannuto) [reply tools/stack_analysis.sh:29]:
```suggestion
# Check if main is the largest stack frame
largest=`echo "$frames" | sort -n -r | head -1 | grep ' main'`
largest_ret_val=$?

# If it is, print a warning.
if [ $largest_ret_val -eq 0 ]; then
    echo "   ${bold}WARNING! main is the largest stack frame!${normal}"
    echo "   See https://github.com/tock/tock/issues/2425 for an explanation of"
    echo "   why this is an issue, and https://github.com/tock/tock/pull/2715 for"
    echo "   an example of how to fix it."
```

**Comment by bradjc** [comment]
This tool has only existed for 138 days. Are you aware of any downstream scripts that process the output?

**Comment by hudson-ayers** [comment]
I am not aware of anyone processing the output of this script downstream, and downstream users are welcome to vendor a copy of the old version if they want the old output format. I do not think that every tool committed to the Tock repository should be locked into its original format, when I published this originally it was just a quick way to let others experiment with solutions to our stack size woes.

**Review by hudson-ayers — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]

**Review by alistair23 — APPROVED** [review]

**Comment by bradjc** [comment]
ping

**Comment by ppannuto** [comment]
bors r+

Phil's point about output stability is well-taken, particularly for things that are 'promoted' (i.e. this isn't just a script in the `tools/` directory, it's advertised via the primary `make` interface). That said, I think the argument that this is a very new tool still figuring out its interface and output holds here. However, that becomes less true after this is bundled as a new `make`-accessible tool with 2.0.

**Comment by hudson-ayers** [comment]
bors retry

**Pull request merged**